### PR TITLE
InfluxDB: Interpolate variables based on their type

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

When we have a variable with a value `8.1` and another variable with value `https://aaaa-aa-aaa.bbb.ccc.ddd:8443/ggggg` we want them interpolated properly. For `8.1` we don't want to escape the dot. For the URL we want it to be escaped. This PR is handling that situation so floating point numbers and URLs (a URL is just an example here. it might be anything you want to be escaped) will be treated differently. 

The interpolation of the template variables was changed here https://github.com/grafana/grafana/pull/65875
This PR is not only fixing that part of the code but also handling all interpolations.

**Why do we need this feature?**

A better variable interpolation

**Who is this feature for?**

InfluxDB users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/67810

**Special notes for your reviewer:**

### How to test it?
<details>

<summary>dashboard.json</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 1165,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "datasource": {
        "type": "influxdb",
        "uid": "q5orKAR4k"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisBorderShow": false,
            "axisCenteredZero": false,
            "axisColorMode": "text",
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "insertNulls": false,
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 1,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom",
          "showLegend": true
        },
        "tooltip": {
          "mode": "single",
          "sort": "none"
        }
      },
      "targets": [
        {
          "datasource": {
            "type": "influxdb",
            "uid": "q5orKAR4k"
          },
          "query": "SELECT sum(\"piece_count\") FROM \"pdata\" WHERE diameter <= $maxSED AND agent_url =~ /^$var1$/",
          "rawQuery": true,
          "refId": "A",
          "resultFormat": "time_series"
        }
      ],
      "title": "bidsolutions escalation 5454",
      "type": "timeseries"
    }
  ],
  "refresh": "",
  "schemaVersion": 38,
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "selected": false,
          "text": "None",
          "value": ""
        },
        "datasource": {
          "type": "influxdb",
          "uid": "vA4bkHenk"
        },
        "definition": "SHOW TAG VALUES WITH KEY = \"agent_url\" WHERE agent_url =~ /^$var1$/",
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "query0",
        "options": [],
        "query": "SHOW TAG VALUES WITH KEY = \"agent_url\" WHERE agent_url =~ /^$var1$/",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "type": "query"
      },
      {
        "current": {
          "selected": false,
          "text": "https://aaaa-aa-aaa.bbb.ccc.ddd:8443/ggggg",
          "value": "https://aaaa-aa-aaa.bbb.ccc.ddd:8443/ggggg"
        },
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "var1",
        "options": [
          {
            "selected": true,
            "text": "https://aaaa-aa-aaa.bbb.ccc.ddd:8443/ggggg",
            "value": "https://aaaa-aa-aaa.bbb.ccc.ddd:8443/ggggg"
          }
        ],
        "query": "https://aaaa-aa-aaa.bbb.ccc.ddd:8443/ggggg",
        "queryValue": "",
        "skipUrlSync": false,
        "type": "custom"
      },
      {
        "description": "8.1",
        "hide": 2,
        "name": "maxSED",
        "query": "8.1",
        "skipUrlSync": false,
        "type": "constant"
      },
      {
        "current": {
          "isNone": true,
          "selected": false,
          "text": "None",
          "value": ""
        },
        "datasource": {
          "type": "influxdb",
          "uid": "vA4bkHenk"
        },
        "definition": "SELECT sum(\"pcount\") FROM \"pdata\" WHERE diameter <= $maxSED",
        "hide": 0,
        "includeAll": false,
        "multi": false,
        "name": "query1",
        "options": [],
        "query": "SELECT sum(\"pcount\") FROM \"pdata\" WHERE diameter <= $maxSED",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "type": "query"
      },
      {
        "current": {
          "selected": true,
          "text": "9.1",
          "value": "9.1"
        },
        "hide": 0,
        "name": "maxSED2",
        "options": [
          {
            "selected": true,
            "text": "9.1",
            "value": "9.1"
          }
        ],
        "query": "9.1",
        "skipUrlSync": false,
        "type": "textbox"
      }
    ]
  },
  "time": {
    "from": "now-30m",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "InfluxQL Variable Investigation",
  "uid": "e2449f40-9795-4f13-a00a-e6193ca80dfd",
  "version": 68,
  "weekStart": ""
}
```
</details>

- Have influxdb dev environment up and running
- Use the `dashboard.json` above to test all this
- Without this PR not all template queries return 200 and panel query fails.
- Switch to this branch
- Template variable queries should return 200 and panel query should return 200

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/plugins/datasource/influxdb/datasource.ts [hunk 1]
```diff
@@ -802,3 +814,15 @@ function timeSeriesToDataFrame(timeSeries: TimeSeries): DataFrame {
     length: values.length,
   };
 }
+
+export function influxRegularEscape(value: any) {
```
Comment (itsmylife) [inline public/app/plugins/datasource/influxdb/datasource.ts:818]: any idea to not use `any` here @joshhunt ?

Comment (joshhunt) [reply public/app/plugins/datasource/influxdb/datasource.ts:818]: grafana data exports `TypedVariableModel` which can be narrowed down with either `if ("multi" in variable)` checks, or `if (variable.type === "query")`

Comment (itsmylife) [reply public/app/plugins/datasource/influxdb/datasource.ts]: I've double checked the types that `templateSrc.replace` needs and I ended up doing this: https://github.com/grafana/grafana/pull/75527/commits/eec6b90f5f5e6bf9caeb904f465f792603dff3eb

Comment (joshhunt) [reply public/app/plugins/datasource/influxdb/datasource.ts]: should this be imported from scenes? feels weird imho

Comment (itsmylife) [reply public/app/plugins/datasource/influxdb/datasource.ts]: I just used the same logic from prometheus https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/datasource.tsx#L335-L352
Where is this in scenes?

Comment (joshhunt) [reply public/app/plugins/datasource/influxdb/datasource.ts]: you wrote `import { CustomFormatterVariable } from '@grafana/scenes';` in your PR. It feels weird to me that this type is being imported from @grafana/scenes, but i see template_srv does so 🤷🏻

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 2]
```diff
@@ -1158,6 +1158,32 @@ The client ID to use for user-assigned managed identity.
...
+
+Tenant ID of the Azure AD Workload Identity.
+
+Allows to override default tenant ID of the Azure AD identity associated with the Kubernetes service account.
+
+### workload_identity_client_id
+
+Client ID of the Azure AD Workload Identity.
+
+Allows to override default client ID of the Azure AD identity associated with the Kubernetes service account.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1179]:
```suggestion
Allows you to override the default client ID of the Azure AD identity associated with the Kubernetes service account.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 3]
```diff
@@ -1158,6 +1158,32 @@ The client ID to use for user-assigned managed identity.
...
+
+For more documentation on Azure AD Workload Identity, review [Azure AD Workload Identity](https://azure.github.io/azure-workload-identity/docs/) documentation.
+
+Disabled by default, needs to be explicitly enabled.
+
+### workload_identity_tenant_id
+
+Tenant ID of the Azure AD Workload Identity.
+
+Allows to override default tenant ID of the Azure AD identity associated with the Kubernetes service account.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1173]:
```suggestion
Allows you to override the default tenant ID of the Azure AD identity associated with the Kubernetes service account.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 4]
```diff
@@ -1158,6 +1158,32 @@ The client ID to use for user-assigned managed identity.
 
 Should be set for user-assigned identity and should be empty for system-assigned identity.
 
+### workload_identity_enabled
+
+Specifies whether Azure AD Workload Identity authentication should be enabled in datasources that support it.
+
+For more documentation on Azure AD Workload Identity, review [Azure AD Workload Identity](https://azure.github.io/azure-workload-identity/docs/) documentation.
+
+Disabled by default, needs to be explicitly enabled.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1167]:
```suggestion
Disabled by default, it needs to be explicitly enabled.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 5]
```diff
@@ -1158,6 +1158,32 @@ The client ID to use for user-assigned managed identity.
...
+
+Client ID of the Azure AD Workload Identity.
+
+Allows to override default client ID of the Azure AD identity associated with the Kubernetes service account.
+
+### workload_identity_token_file
+
+Custom path to token file for the Azure AD Workload Identity.
+
+Allows to set a custom path to the projected service account token file.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1185]:
```suggestion
Allows you to set a custom path to the projected service account token file.
```

File: docs/sources/setup-grafana/configure-grafana/_index.md [hunk 6]
```diff
@@ -1158,6 +1158,32 @@ The client ID to use for user-assigned managed identity.
 
 Should be set for user-assigned identity and should be empty for system-assigned identity.
 
+### workload_identity_enabled
+
+Specifies whether Azure AD Workload Identity authentication should be enabled in datasources that support it.
+
+For more documentation on Azure AD Workload Identity, review [Azure AD Workload Identity](https://azure.github.io/azure-workload-identity/docs/) documentation.
```
Comment (chri2547) [inline docs/sources/setup-grafana/configure-grafana/_index.md:1165]:
```suggestion
For more documentation on Azure AD Workload Identity, refer to [Azure AD Workload Identity](https://azure.github.io/azure-workload-identity/docs/).
```

**Comment by itsmylife** [comment]
I made a mistake with rebasing 🤦‍♂️

**Comment by itsmylife** [comment]
New PR https://github.com/grafana/grafana/pull/75653

**Pull request closed without merging**